packages/analytics/amplify_analytics_pinpoint_dart/test/string_database_test.dart

```diff
@@ -0,0 +1,197 @@
+        await db.addString((startValue + i).toString());
+      }
+    }
+
+    /// Reload db instance
+    Future<void> refreshDb() async {
+      db = DartStringDatabase(null);
+    }
+
+    void validateElements(
+      Iterable<StringDatabaseElement> readElements,
+      List<String> targetValues,
+    ) {
+      expect(readElements.length, targetValues.length);
+      for (var i = 0; i < readElements.length; i++) {
+        expect(
+          reason: 'Comparing element $i of readElements vs targetValues',
+          readElements.elementAt(i).string,
+          targetValues[i],
+        );
+      }
+    }
+
+    Future<Iterable<StringDatabaseElement>> readAndValidateValues(
+      int readCount, {
+      int startValue = 0,
+      bool checkExactCount = true,
+    }) async {
+      final readElements = await db.getStrings(readCount);
+      if (checkExactCount) {
+        expect(readElements.length, readCount);
+      }
+      var currentIndex = startValue;
+      for (final readValue in readElements) {
+        expect(readValue.string, currentIndex.toString());
+        currentIndex++;
+      }
+      return readElements;
+    }
+
+    List<int> convertReadElementsToIds(
+      Iterable<StringDatabaseElement> readElements,
+    ) {
+      final ids = <int>[];
+      for (final element in readElements) {
+        ids.add(element.id);
+      }
+      return ids;
+    }
+
+    Future<void> deleteAndValidateValues({
+      required int writeCount,
+      required List<int> idsToDelete,
+    }) async {
+      await writeValues(writeCount);
+
+      final readElements = await readAndValidateValues(writeCount);
+      final readList = readElements.toList();
+
+      // Because startIndex is not cleared in IndexedDb
+      final shiftedIdsToDelete = <int>[];
+      for (var i = 0; i < idsToDelete.length; i++) {
+        shiftedIdsToDelete.add(readList[idsToDelete[i]].id);
+      }
+
+      await db.deleteStrings(shiftedIdsToDelete);
+
+      final newReadElements = await db.getStrings(writeCount);
+
+      final expectedList = <String>[];
+      for (var i = 0; i < writeCount; i++) {
+        if (!idsToDelete.contains(i)) {
+          expectedList.add(i.toString());
+        }
+      }
+      validateElements(newReadElements, expectedList);
+    }
+
+    setUp(() async {
+      await refreshDb();
+    });
+
+    tearDown(() async {
+      await db.clear();
+    });
+
+    test('Strings stored', () async {
+      await writeValues(10);
+      await readAndValidateValues(5);
+    });
+
+    test('getStrings can return all stored strings', () async {
+      await writeValues(10);
+      await readAndValidateValues(100, checkExactCount: false);
+    });
+
+    test('Delete strings', () async {
+      await deleteAndValidateValues(
+        writeCount: 10,
+        idsToDelete: [0, 1, 2, 3],
+      );
+
+      await db.clear();
+
+      await deleteAndValidateValues(
+        writeCount: 10,
+        idsToDelete: [6, 7, 8, 9],
+      );
+    });
```

Comment:
The code in these tests is _very_ DRY, which usually is a good thing, but sometimes in tests it is not very desirable as it reduces the readability. Take a look at https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests

Good tests should be easy to ready, and clearly explain the expected behavior of the code, even if that results in some duplication.

Reply:
Oh wow I didn't know that before but it makes a late of sense.  Rewrote the unit test to be DAMP instead of DRY.